Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp

```diff
@@ -0,0 +1,375 @@
+  {
+    default_host_controls = default_host_controls_json.get<picojson::object>();
+  }
+
+  const auto output_textures = output_textures_json.get<picojson::object>();
+  for (auto& [name, data] : output_textures)
+  {
+    DynamicInputTextureData texture_data;
+    texture_data.m_hires_texture_name = name;
+
+    // Required fields
+    const picojson::value& colorkey_image = data.get("colorkey_image");
+    const picojson::value& fallback_image = data.get("fallback_image");
+    const picojson::value& emulated_controls = data.get("emulated_controls");
+
+    if (!colorkey_image.is<std::string>() || !emulated_controls.is<picojson::object>() ||
+        !fallback_image.is<std::string>())
+    {
+      ERROR_LOG(VIDEO,
+                "Failed to load dynamic input json file '%s' because required fields "
+                "'colorkey_image', 'fallback_image', or 'emulated_controls' are either "
+                "missing or the incorrect type",
+                json_file.c_str());
+      m_valid = false;
+      return;
+    }
+
+    texture_data.m_color_key_image_name = colorkey_image.to_str();
+    texture_data.m_fallback_image_name = fallback_image.to_str();
+    texture_data.m_preserve_aspect_ratio = preserve_aspect_ratio;
+
+    SplitPath(json_file, &m_base_path, nullptr, nullptr);
+
+    const std::string color_key_image_full_path = m_base_path + texture_data.m_color_key_image_name;
+    if (!File::Exists(color_key_image_full_path))
+    {
+      ERROR_LOG(VIDEO,
+                "Failed to load dynamic input json file '%s' because the color key image '%s' "
+                "could not be loaded",
+                json_file.c_str(), color_key_image_full_path.c_str());
+      m_valid = false;
+      return;
+    }
+
+    const std::string fallback_image_full_path = m_base_path + texture_data.m_fallback_image_name;
+    if (!File::Exists(fallback_image_full_path))
+    {
+      ERROR_LOG(VIDEO,
+                "Failed to load dynamic input json file '%s' because the fallback image '%s' "
+                "could not be loaded",
+                json_file.c_str(), fallback_image_full_path.c_str());
+      m_valid = false;
+      return;
+    }
+
+    const auto& emulated_controls_json = emulated_controls.get<picojson::object>();
+    for (auto& [emulated_controller_name, map] : emulated_controls_json)
+    {
+      if (!map.is<picojson::object>())
+      {
+        ERROR_LOG(VIDEO,
+                  "Failed to load dynamic input json file '%s' because 'emulated_controls' "
+                  "map key '%s' is incorrect type ",
+                  json_file.c_str(), emulated_controller_name.c_str());
+        m_valid = false;
+        return;
+      }
+      auto& key_to_color = texture_data.m_emulated_controllers[emulated_controller_name];
+      for (auto& [emulated_control, color] : map.get<picojson::object>())
+      {
+        const std::string color_as_string = color.to_str();
+
+        bool color_code_error = false;
```

Comment:
Could you extract the whole color parsing logic into its own function please? The function could return a `std::optional<int>` or even `std::optional<DynamicInputTextureData::ColorKey>`. I think that'd make this function easier to read.

Reply:
Good idea!